KAFKA-9432 automated protocol for DescribeConfigs

tombentley opened the pull request:
Change the DescribeConfigs request and response to use the generated protocol messages.

tombentley committed:
KAFKA-9432 automated protocol for DescribeConfigs

tombentley commented:
@mimaison here's another one, grateful if you could review.

tombentley commented:
Rebased for conflict.

tombentley commented:
Rebased for conflict.

tombentley commented:
Rebased for conflict.

tombentley committed:
rebase

tombentley commented:
Rebased for conflict.

mimaison commented:
retest this please

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1912,53 +1915,51 @@ public DescribeConfigsResult describeConfigs(Collection<ConfigResource> configRe
                 DescribeConfigsRequest.Builder createRequest(int timeoutMs) {
-                    return new DescribeConfigsRequest.Builder(unifiedRequestResources)
-                            .includeSynonyms(options.includeSynonyms());
+                    return new DescribeConfigsRequest.Builder(unifiedRequestResources.setIncludeSynoyms(options.includeSynonyms()));
```
I wonder if it would read better if `unifiedRequestResources` was instead a `List<DescribeConfigsResource>` and we were creating the `DescribeConfigsRequestData` object and calling all its setters here. WDYT?

mimaison commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsRequest.java:
```diff
@@ -92,85 +48,43 @@ public Builder(Collection<ConfigResource> resources) {
+        return new DescribeConfigsResponse(new DescribeConfigsResponseData()
+                .setThrottleTimeMs(throttleTimeMs)
+                .setResults(data.resources().stream().map(r -> {
+                    Errors error = Errors.forException(e);
```
Can we move that line to the top?

mimaison commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -220,145 +153,66 @@ public ConfigSource source() {
+        this.data = new DescribeConfigsResponseData(struct, version);
+        if (version == 0) {
+            for (DescribeConfigsResult result : data.results()) {
+                for (DescribeConfigsResponseData.DescribeConfigsResourceResult x : result.configs()) {
```
Let's give that variable a better name, what about `config`?

mimaison commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -1845,34 +1857,59 @@ private DeleteAclsResponse createDeleteAclsResponse() {
+                new DescribeConfigsResult()
+                        .setResourceType(ConfigResource.Type.BROKER.id())
+                        .setResourceName("0")
+                        .setErrorCode(ApiError.NONE.error().code())
```
We can use `Errors.NONE.code()` directly. Same below

mimaison commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -347,59 +346,62 @@ class AdminManager(val config: KafkaConfig,
 
         val configEntries = filteredConfigPairs.map { case (name, value) => createConfigEntry(name, value) }
-        new DescribeConfigsResponse.Config(ApiError.NONE, configEntries.asJava)
+        new DescribeConfigsResponseData.DescribeConfigsResult().setErrorCode(ApiError.NONE.error().code())
```
We can use `Errors.NONE.code` directly

mimaison commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -707,11 +718,12 @@ class AdminManager(val config: KafkaConfig,
     }
     val allSynonyms = configSynonyms(name, allNames, isSensitive)
-        .filter(perBrokerConfig || _.source == ConfigSource.DYNAMIC_DEFAULT_BROKER_CONFIG)
+        .filter(perBrokerConfig || _.source == ConfigSource.DYNAMIC_DEFAULT_BROKER_CONFIG.id())
```
No need for the brackets after `id`

mimaison commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2561,25 +2561,32 @@ class KafkaApis(val requestChannel: RequestChannel,
     val unauthorizedConfigs = unauthorizedResources.map { resource =>
-      val error = configsAuthorizationApiError(resource)
-      resource -> new DescribeConfigsResponse.Config(error, util.Collections.emptyList[DescribeConfigsResponse.ConfigEntry])
+      val error = ConfigResource.Type.forId(resource.resourceType()) match {
```
We can remove the brackets here. Same a few lines below after `resourceName`

mimaison commented on clients/src/main/resources/common/message/DescribeConfigsRequest.json:
```diff
@@ -17,7 +17,7 @@
   "type": "request",
   "name": "DescribeConfigsRequest",
-  // Version 1 adds IncludeSynoyms.
+  // Version 1 adds IncludeSynonyms.
```
Let's also fix the field name

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2017,6 +2016,18 @@ void handleFailure(Throwable throwable) {
         return new DescribeConfigsResult(allFutures);
     }
 
+    private Config describeConfigResult(DescribeConfigsResponseData.DescribeConfigsResult describeConfigsResult) {
```
It looks like we are not calling `configSynonyms()` anymore, so we can remove that private method

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1912,53 +1915,51 @@ public DescribeConfigsResult describeConfigs(Collection<ConfigResource> configRe
-                                    configEntry.isSensitive(), configEntry.isReadOnly(),
-                                    configSynonyms(configEntry)));
+
+                        for (DescribeConfigsResponseData.DescribeConfigsResult resultData : response.data().results()) {
```
We are not using `resultData` in this loop. Is that expected?

mimaison commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsRequest.java:
```diff
@@ -17,70 +17,26 @@
-
-        public Builder(Collection<ConfigResource> resources) {
-            this(toResourceToConfigNames(resources));
+            this.data = data;
```
As we remove the other constructors, `toResourceToConfigNames` is not called anymore. Let's delete it

mimaison commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -174,19 +97,25 @@ public boolean isReadOnly() {
-        STATIC_BROKER_CONFIG((byte) 4),
-        DEFAULT_CONFIG((byte) 5),
-        DYNAMIC_BROKER_LOGGER_CONFIG((byte) 6);
+        UNKNOWN_CONFIG((byte) 0, org.apache.kafka.clients.admin.ConfigEntry.ConfigSource.UNKNOWN),
```
It's unfortunate the public enum cannot do the mapping and we have to keep a copy here. Could we maybe rename this one so this looks less crazy? WDYT?

mimaison reviewed (changes requested):
@tombentley Thanks for the PR. I've made a first pass and left a bunch of comments

tombentley committed:
Review comments

tombentley commented:
@mimaison I've fixed those, if you want to make a 2nd pass. I guess I should add a test to `KafkaAdminClientTest` too.

mimaison commented:
retest this please

mimaison commented:
Let's get https://github.com/apache/kafka/pull/8311 merged first and then rebase this one on top

tombentley committed:
Fix rebase

tombentley committed:
Better test

tombentley committed:
Use single request when getting BROKER & BROKER_LOGGER from same broker

tombentley commented:
@mimaison I've rebased and added some more tests in KafkaAdminClientTest. In so doing I noticed that `describeConfigs()` would require two requests to each broker when both `BROKER` and `BROKER_LOGGER` config resources for a broker were requested. I've fixed that, but `alterConfigs()` and `incrementalAlterConfigs()` do the same thing. WDYT about changing them too?

tombentley commented:
This PR will also fix [KAFKA-6962](https://issues.apache.org/jira/browse/KAFKA-6962)

mimaison commented:
retest this please

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1916,129 +1917,96 @@ void handleFailure(Throwable throwable) {
-                        for (DescribeConfigsResponse.ConfigEntry configEntry : config.entries()) {
-                            configEntries.add(new ConfigEntry(configEntry.name(), configEntry.value(),
-                                configSource(configEntry.source()), configEntry.isSensitive(), configEntry.isReadOnly(),
-                                configSynonyms(configEntry), configType(configEntry.type()), configEntry.documentation()));
```
It looks like `configType()` is not used anymore. Can we delete it?

mimaison commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2568,25 +2568,32 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+      new DescribeConfigsResponseData.DescribeConfigsResult().setErrorCode(error.code)
+        .setErrorMessage(error.message)
+        .setConfigs(util.Collections.emptyList[DescribeConfigsResponseData.DescribeConfigsResourceResult])
```
We can drop `util` here, we are importing `import java.util.{Collections, Optional}`

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1916,129 +1917,96 @@ void handleFailure(Throwable throwable) {
+            brokerFutures.compute(broker, (key, value) -> {
+                if (value == null) {
+                    // Only BROKER and BROKER_LOGGER configs are broker-specific
+                    value = new HashMap<>(broker != null ? 2 : 16);
```
I wonder if we can just get a Map with the default size. I don't expect this code path to be very hot

mimaison reviewed (changes requested):
Thanks for the updates. It looks good. I just left a few minor comments.

tombentley committed:
Review comments

tombentley commented:
@mimaison done.

ableegoldman commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -293,158 +208,66 @@ public ConfigSource source() {
-    private final Map<ConfigResource, Config> configs;
+    private final DescribeConfigsResponseData data;
 
-    public DescribeConfigsResponse(int throttleTimeMs, Map<ConfigResource, Config> configs) {
-        this.throttleTimeMs = throttleTimeMs;
-        this.configs = Objects.requireNonNull(configs, "configs");
```
Hey @mimaison @tombentley, removing this constructor seems to have broken the build (it was used in `TopicAdminTest`). Can we get a hotfix for this to unblock other PRs?
edit: nevermind, a PR was just opened for it ([pull/8866](https://github.com/apache/kafka/pull/8866))

tombentley commented:
@ableegoldman sorry about that. A fix has now been committed to trunk. Thanks to @chia7712.

omkreddy commented:
@tombentley After this change, `adminClient.describeConnfigs()` is not backward compatible with older kafka broker versions.  It seems, older brokers are receiving empty configNames [here](https://github.com/apache/kafka/blob/2.5/core/src/main/scala/kafka/server/AdminManager.scala#L355). Older brokers does not handle empty configKeys.  Due to this old brokers are returning empty config list.  We may need to explicitly set `setConfigurationKeys(null)` [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L1952).

We also need to update ClientCompatibilityTest, to catch these kind of errors.
https://github.com/apache/kafka/blob/trunk/tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java
https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/client/client_compatibility_features_test.py

cc @mimaison

tombentley commented:
@omkreddy sorry about that, I was away. I see that you fixed it in the meantime.

tombentley commented:
@omkreddy, also thanks for the pointer to that compatibility test, I wasn't previously aware of it, but I'll use it now that I am.
